engine/consensus/matching/engine.go

```diff
@@ -1045,27 +1110,60 @@ func (e *Engine) requestPendingReceipts() (int, error) {
 		// get the block header at this height (should not error as heights are finalized)
 		header, err := e.headersDB.ByHeight(height)
 		if err != nil {
-			return 0, fmt.Errorf("could not get header (height=%d): %w", height, err)
+			return 0, 0, fmt.Errorf("could not get header (height=%d): %w", height, err)
 		}
 
 		// check if we have an result for the block at this height
 		blockID := header.ID()
-		if _, ok := knownResultForBlock[blockID]; !ok {
-			missingBlocksOrderedByHeight = append(missingBlocksOrderedByHeight, blockID)
+
+		_, ok := knownResultForBlock[blockID]
+		if ok {
+			continue
+		}
+
+		// since the index is only added when the block which includes the receipts
+		// get finalized, so the returned receipts must be from finalized blocks.
+		// Therefore, the return receipts must be incoporated receipts, which
+		// are safe to be added to the mempool
+		receipts, err := e.receiptsDB.ByBlockIDAllExecutionReceipts(blockID)
+		if err != nil {
+			return 0, 0, fmt.Errorf("could not get receipts by block ID: %v, %w", blockID, err)
+		}
+
+		if len(receipts) > 0 {
+			for _, receipt := range receipts {
+				_, err = e.receipts.AddReceipt(receipt, header)
+				if err != nil {
+					return 0, 0, fmt.Errorf("could not add receipt to receipts mempool %v, %w", receipt.ID(), err)
+				}
+
+				_, err = e.incorporatedResults.Add(
+					flow.NewIncorporatedResult(
+						receipt.ExecutionResult.BlockID,
+						&receipt.ExecutionResult,
+					),
+				)
+
+				if err != nil {
+					return 0, 0, fmt.Errorf("could not add result to incorporated results mempool %v, %w", receipt.ID(), err)
+				}
+			}
+			continue
+		}
+
+		missingBlocksOrderedByHeight = append(missingBlocksOrderedByHeight, blockID)
+		if firstMissingHeight == 0 {
+			firstMissingHeight = height
 		}
+
 	}
 
 	// request missing execution results, if sealed height is low enough
-	if len(missingBlocksOrderedByHeight) > 0 {
-		log.Debug().
-			Int("finalized_blocks_without_result", len(missingBlocksOrderedByHeight)).
-			Msg("requesting receipts")
-		for _, blockID := range missingBlocksOrderedByHeight {
-			e.receiptRequester.EntityByID(blockID, filter.Any)
-		}
+	for _, blockID := range missingBlocksOrderedByHeight {
+		e.receiptRequester.EntityByID(blockID, filter.Any)
```

Comment:
receipts should be fetched with `Query`. Probably not for this branch, but just keep in mind